diff --git a/.clang-format b/.clang-format
index 86a5093..5e609a8 100644
--- a/.clang-format
+++ b/.clang-format
@@ -1,6 +1,6 @@
 ---
-Language:        Cpp
-BasedOnStyle:  Google
+Language: Cpp
+BasedOnStyle: Google
 DerivePointerAlignment: false
 PointerAlignment: Left
 AlignConsecutiveAssignments: false
diff --git a/src/lp_interface/BUILD.bazel b/src/lp_interface/BUILD.bazel
index 79b923b..178b9f8 100644
--- a/src/lp_interface/BUILD.bazel
+++ b/src/lp_interface/BUILD.bazel
@@ -70,6 +70,8 @@ cc_library(
     name = "lpi_factory",
     srcs = ["lpi_factory.cc"],
     hdrs = ["lpi_factory.h"],
+    # Required for proper linking with soplex.
+    linkstatic = True,
     deps = [
         ":lpi",
         ":lpi_glop",
diff --git a/src/lp_interface/lpi_soplex.cc b/src/lp_interface/lpi_soplex.cc
index 7a584e1..745733b 100644
--- a/src/lp_interface/lpi_soplex.cc
+++ b/src/lp_interface/lpi_soplex.cc
@@ -656,7 +656,7 @@ LpSoplexInterface::DeleteRowSet(
 
   SOPLEX_TRY(spx_->removeRowsReal(int_deletion_status.data()));
 
-  // Note: We cannot use the same vector for the soplex call above and the row
+  // Note: We cannot use the same vector for the Soplex call above and the row
   // mapping, even though RowIndex internally consists of an int. In addition to
   // being bad practice to depend on a class' internals, the compiler may
   // introduce additional padding meaning RowIndex and int don't have identical
